presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,454 @@
+    }
+
+    private List<ElasticsearchColumn> buildColumns(ElasticsearchTableDescription tableDescription)
+    {
+        List<ElasticsearchColumn> columns = new ArrayList<>();
+        TransportClient client = clients.get(tableDescription.getClusterName());
+        verify(client != null, "client is null");
+        for (String index : getIndices(tableDescription)) {
+            GetMappingsRequest mappingsRequest = new GetMappingsRequest().types(tableDescription.getType());
+
+            if (!isNullOrEmpty(index)) {
+                mappingsRequest.indices(index);
+            }
+            ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = getMappings(client, mappingsRequest);
+
+            Iterator<String> indexIterator = mappings.keysIt();
+            while (indexIterator.hasNext()) {
+                // TODO use io.airlift.json.JsonCodec
+                MappingMetaData mappingMetaData = mappings.get(indexIterator.next()).get(tableDescription.getType());
+                JsonNode rootNode;
+                try {
+                    rootNode = objecMapper.readTree(mappingMetaData.source().uncompressed());
+                }
+                catch (IOException e) {
+                    throw new PrestoException(ELASTICSEARCH_CORRUPTED_MAPPING_METADATA, e);
+                }
+                // parse field mapping JSON: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-get-field-mapping.html
+                JsonNode mappingNode = rootNode.get(tableDescription.getType());
+                JsonNode propertiesNode = mappingNode.get("properties");
+
+                List<String> lists = new ArrayList<>();
+                JsonNode metaNode = mappingNode.get("_meta");
+                if (metaNode != null) {
+                    JsonNode arrayNode = metaNode.get("lists");
+                    if (arrayNode != null && arrayNode.isArray()) {
+                        ArrayNode arrays = (ArrayNode) arrayNode;
+                        for (int i = 0; i < arrays.size(); i++) {
+                            lists.add(arrays.get(i).textValue());
+                        }
+                    }
+                }
+                populateColumns(propertiesNode, lists, columns);
+            }
+        }
+        return columns;
+    }
+
+    private ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> getMappings(TransportClient client, GetMappingsRequest request)
+    {
+        try {
+            return retry()
+                    .maxAttempts(maxAttempts)
+                    .exponentialBackoff(maxRetryTime)
+                    .run("getMappings", () -> client.admin()
+                            .indices()
+                            .getMappings(request)
+                            .actionGet(requestTimeout.toMillis())
+                            .getMappings());
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
can we wrap this try-catch block as method and then just call `call("getMappings", () -> client...);`?